### PR TITLE
Add support for compose CI messages

### PR DESCRIPTION
**Pull request by hluk**
---
The new message format is still in review (FACTORY-2536) -- maybe it's already finished since we didn't get any complaints.

The fake message for tests are missing some fields - I basically just copy-pasted it from other JSON and set only the required fields and obfuscated some real values.

I'll update ansible repo later.

@danc86, please review.

### CODE REVIEW

**Comment by danc86** [comment]
I think we will need to add 'arch' now.

Also maybe let's just rush this in just yet, until all outstanding questions about the message format are settled.

**Comment by danc86** [comment]
Uhh... meant to say: let's *not* rush this in just yet :-)

File: resultsdbupdater/utils.py [hunk 1]
```diff
@@ -236,25 +227,49 @@ def handle_ci_umb(msg):
...
-        'system_provider': system.get('provider'),
-        'ci_name': msg_body_ci.get('name'),
-        'ci_url': msg_body_ci.get('url'),
-        'ci_environment': msg_body_ci.get('environment'),
-        'ci_team': msg_body_ci.get('team'),
-        'ci_irc': msg_body_ci.get('irc'),
-        'ci_email': msg_body_ci.get('email'),
-    }
+    if item_type == "compose":
+        result_data = {
```
Comment (mprahl) [inline resultsdbupdater/utils.py]: Optional: You could define `result_data` outside the if statement with the common key value pairs, and then update the dict in the if statements

Comment (hluk) [reply resultsdbupdater/utils.py]: Tried that but it looks bit easier to read and change this way.

**Comment by hluk** [comment]
I've added commit to properly handle messages with `*.error` topics.

**Comment by hluk** [comment]
I've added another small fix to use "productmd.compose.id" attribute name (used by OpenQA) instead of "compose_id". Otherwise this discrepancy needs to be resolved in Greenwave.

**Comment by danc86** [comment]
:+1: to these changes

File: resultsdbupdater/utils.py [hunk 2]
```diff
@@ -236,25 +235,58 @@ def handle_ci_umb(msg):
...
+
+                ('type', item_type),
+                ('productmd.compose.id', msg_body['artifact']['compose_id']),
+
+                ('system_provider', system['provider']),
+                ('system_architecture', system['architecture']),
+                ('system_variant', system.get('variant')),
+
+                ('category', msg_body.get('category')),
+            ) if value is not None
```
Comment (mprahl) [inline resultsdbupdater/utils.py]: Optional: I'm wondering if it's better to have consistency and just put null values for keys that aren't present in the message but should be. What do you think?

File: resultsdbupdater/utils.py [hunk 3]
```diff
@@ -236,25 +235,58 @@ def handle_ci_umb(msg):
...
-        'system_os': system.get('os'),
-        'system_provider': system.get('provider'),
-        'ci_name': msg_body_ci.get('name'),
-        'ci_url': msg_body_ci.get('url'),
-        'ci_environment': msg_body_ci.get('environment'),
-        'ci_team': msg_body_ci.get('team'),
-        'ci_irc': msg_body_ci.get('irc'),
-        'ci_email': msg_body_ci.get('email'),
-    }
+    if item_type == "compose":
```
Comment (mprahl) [inline resultsdbupdater/utils.py]: For consistency, it'd be nice to use single quotes since the rest of file does.

File: tests/fake_messages/pipeline_failure_message.json [hunk 4]
```diff
@@ -21,7 +21,6 @@
   "body": {
     "msg": {
       "category": "functional",
-      "status": "FAILED",
```
Comment (mprahl) [inline tests/fake_messages/pipeline_failure_message.json]: Why was this line removed?

File: tests/test_consumer.py [hunk 5]
```diff
@@ -656,3 +656,55 @@ def test_full_consume_osci_example_2(mock_get_session, namespace, expected):
...
+    with open(fake_msg_path) as fake_msg_file:
+        fake_msg = json.load(fake_msg_file)
+
+    assert consumer.consume(fake_msg) is True
+    # Verify the URLs called
+    assert mock_requests.post.call_args_list[0][0][0] == \
+        'https://resultsdb.domain.local/api/v2.0/results'
+    # Verify the post data
+    assert mock_requests.post.call_count == 1
+    url = "https://rtt-jenkins/job/compose-RHEL-X.0-rel-eng-tier2-acceptance/1/"
```
Comment (mprahl) [inline tests/test_consumer.py]: It'd be nice to single quotes since the rest of the project does.

File: tests/test_consumer.py [hunk 6]
```diff
@@ -656,3 +656,55 @@ def test_full_consume_osci_example_2(mock_get_session, namespace, expected):
...
+    expected_data = {
+        "testcase": {
+            "name": "unknown.tier2.functional",
```
Comment (mprahl) [inline tests/test_consumer.py]: You should probably address the fact that we are getting `unknown` here. Are you using an old message that doesn't have `namespace` set?

**Comment by mprahl** [comment]
@danc86, no blockers from me but a few requests and suggestions. Feel free to merge and release if you're in a hurry.

File: resultsdbupdater/utils.py (hunk 2 above)
Comment (danc86) [reply resultsdbupdater/utils.py]: Resultsdb doesn't accept null values for the extra data though, does it? It is a mapping string -> list of string.

File: tests/fake_messages/pipeline_failure_message.json (hunk 4 above)
Comment (hluk) [reply tests/fake_messages/pipeline_failure_message.json]: The is no "status" attribute for these "error" messages (see [Factory 2.0 CI UMB messages](https://docs.google.com/document/d/16L5odC-B4L6iwb9dp8Ry0Xk5Sc49h9KvTHrG86fdfQM/edit#heading=h.gjun9319pfoj)).

File: tests/test_consumer.py (hunk 6 above)
Comment (hluk) [reply tests/test_consumer.py]: Nice catch! Adding the missing (required) "namespace" attribute.

**Comment by hluk** [comment]
@mprahl Thanks for the comments. I've replaced the quotes and added "namespace" attribute to the test message.

**Comment by hluk** [comment]
Can someone merge this? Dan mention doesn't have the permissions, me neither.

Any idea when the changes can be deployed?

**Comment by ralphbean** [comment]
> Can someone merge this? Dan mention doesn't have the permissions, me neither.

I'll add you guys so you have rights.

> Any idea when the changes can be deployed?

Matt and I are the only ones who have cut releases before.  It would be good to get you rights to do this too.  :/

**Comment by ralphbean** [comment]
OK, I tagged a v5.1.0.

**Comment by ralphbean** [comment]
In the future check out the resultsdb-updater package with `rhpkg`.

**Comment by ralphbean** [comment]
5.1.0 rpm is being built in Brew.

@hluk, can you carry the deployment forwards tomorrow?  @mprahl wrote a mojo blogpost titled "Factory 2.0: Deploying a Service in PDI" with some pointers.

**Comment by hluk** [comment]
> can you carry the deployment forwards tomorrow?

I'm on it.
